[ARM][Thumb2] Mark BTI-clearing instructions as scheduling region boundaries

vhscampos opened the pull request:
Following https://github.com/llvm/llvm-project/pull/68313 this patch extends the idea to M-profile PACBTI.

The Machine Scheduler can reorder instructions within a scheduling region depending on the scheduling policy set. If a BTI-clearing instruction happens to partake in one such region, it might be moved around, therefore ending up where it shouldn't.

The solution is to mark all BTI-clearing instructions as scheduling region boundaries. This essentially means that they must not be part of any scheduling region, and as consequence never get moved:

 - PAC
 - PACBTI
 - BTI
 - SG

Note that PAC isn't BTI-clearing, but it's replaced by PACBTI late in the compilation pipeline.

As far as I know, currently it isn't possible to organically obtain code that's susceptible to the bug:

 - Instructions that write to SP are region boundaries. PAC seems to always be followed by the pushing of r12 to the stack, so essentially PAC is always by itself in a scheduling region.
 - CALL_BTI is expanded into a machine instruction bundle. Bundles are unpacked only after the last machine scheduler run. Thus setjmp and BTI can be separated only if someone deliberately run the scheduler once more.
 - The BTI insertion pass is run late in the pipeline, only after the last machine scheduling has run. So once again it can be reordered only if someone deliberately runs the scheduler again.

Nevertheless, one can reasonably argue that we should prevent the bug in spite of the compiler not being able to produce the required conditions for it. If things change, the compiler will be robust against this issue.

The tests written for this are contrived: bogus MIR instructions have been added adjacent to the BTI-clearing instructions in order to have them inside non-trivial scheduling regions.

vhscampos committed:
Mark BTI-clearing instructions as scheduling region boundaries

Following https://github.com/llvm/llvm-project/pull/68313 this patch
extends the idea to M-profile PACBTI.

The Machine Scheduler can reorder instructions within a scheduling
region depending on the scheduling policy set. If a BTI-clearing
instruction happens to partake in one such region, it might be moved
around, therefore ending up where it shouldn't.

The solution is to mark all BTI-clearing instructions as scheduling
region boundaries. This essentially means that they must not be part of
any scheduling region, and as consequence never get moved:

 - PAC
 - PACBTI
 - BTI
 - SG
 - CALL_BTI (pseudo-instruction for setjmp + bti)

Note that PAC isn't BTI-clearing, but it's replaced by PACBTI late in
the compilation pipeline.

As far as I know, currently it isn't possible to organically obtain code
that's susceptible to the bug:

 - Instructions that write to SP are region boundaries. PAC seems to
   always be followed by the pushing of r12 to the stack, so essentially
   PAC is always by itself in a scheduling region.
 - CALL_BTI is expanded into a machine instruction bundle. Bundles are
   unpacked only after the last machine scheduler run. Thus setjmp and
   BTI can be separated only if someone deliberately runs the scheduler
   once more.
 - The BTI insertion pass is run late in the pipeline, only after the
   last machine scheduling has run. So once again it can be reordered
   only if someone deliberately runs the scheduler again.

Nevertheless, one can reasonably argue that we should prevent the bug in
spite of the compiler not being able to produce the required conditions
for it. If things change, the compiler will be robust against this
issue.

The tests written for this are contrived: bogus MIR instructions have
been added adjacent to the BTI-clearing instructions in order to have
them inside non-trivial scheduling regions.

atrosinenko commented on llvm/test/CodeGen/ARM/misched-branch-targets.mir:
```diff
@@ -0,0 +1,181 @@
+# RUN: llc -o - -run-pass=postmisched %s | FileCheck %s
+
+--- |
+  target datalayout = "e-m:e-p:32:32-Fi8-i64:64-v128:64:128-a:0:32-n32-S64"
```
The inline LLVM IR module can probably be simplified by dropping non-essential attributes, as suggested in [TestingGuide](https://llvm.org/docs/TestingGuide.html). [MIR-specific](https://llvm.org/docs/MIRLangRef.html#simplifying-mir-files) documentation also suggests that lots of autogenerated YAML attributes are likely to be safe to remove, but many are dropped already in this test.

I guess the machine functions themselves can be made simpler by dropping `CFI_INSTRUCTION`s and so on, but I'm not sure.

vhscampos replied:
Thanks for the suggestions. I've removed a good chunk of unnecessary info.

atrosinenko reviewed (commented):
A few thoughts/suggestions on the tests (just an opinion):
* the test can probably be further simplified by dropping unrelated machine instructions or replacing them with obvious NOPs for readability
* AFAIK in their inline LLVM IR modules, some mir-based tests use dummy function definitions as placeholders - not corresponding to the actual machine instructions (probably, some properties of the original functions have to be kept)
* it may be worth adding simple test cases for the other BTI-clearing instructions

vhscampos committed:
fixup! Mark BTI-clearing instructions as scheduling region boundaries

Addressed comments.

vhscampos commented on llvm/lib/Target/ARM/ARMBaseInstrInfo.cpp:
```diff
@@ -2088,7 +2088,7 @@ bool ARMBaseInstrInfo::isSchedulingBoundary(const MachineInstr &MI,
     return true;
 
-  return false;
+  return TargetInstrInfo::isSchedulingBoundary(MI, MBB, MF);
```
I've reverted this part of the patch so not to shake things up here.

vhscampos commented:
> * the test can probably be further simplified by dropping unrelated machine instructions or replacing them with obvious NOPs for readability

I've removed one or two extra instructions. I reckon that what's left is important because it's intrinsic to the prolog/epilog codegen of PAC and BTI, although some of it is not necessary to trigger the bug.

> * AFAIK in their inline LLVM IR modules, some mir-based tests use dummy function definitions as placeholders - not corresponding to the actual machine instructions (probably, some properties of the original functions have to be kept)

Good point. Now all the IR definitions are as dummy as possible.

> * it may be worth adding simple test cases for the other BTI-clearing instructions

Added one test for SG and another for PACBTI.

I removed t2CALL_BTI from the list of scheduling boundaries because it's a pseudo-instruction that is transformed into BL + BTI. The pseudo-inst itself can be moved around with no issues. After the expansion, the restrictions must then be enforced.

vhscampos commented:
@atrosinenko Can you please have a look at the latest patch? Thanks

atrosinenko reviewed (approved):
Looks good to me, sorry for long delay. I unresolved a comment that was likely to be resolved unintentionally, though that was a minor issue anyway.

vhscampos committed:
fixup! fixup! Mark BTI-clearing instructions as scheduling region boundaries

vhscampos commented on llvm/test/CodeGen/ARM/misched-branch-targets.mir:
```diff
@@ -0,0 +1,124 @@
+  declare void @longjmp(ptr noundef, i32 noundef) #3
+
+  attributes #0 = { "frame-pointer"="all" "target-cpu"="cortex-m55" "target-features"="+armv8.1-m.main" }
+  attributes #1 = { "frame-pointer"="all" "target-cpu"="cortex-m55" "target-features"="+armv8.1-m.main" }
```
Removed duplicate attribute list. Thanks

vhscampos commented:
Thanks @atrosinenko . I've resolved the attribute list duplication.

zeroomega commented:
We understand the issue now. The `shuffle` option is only available when building LLVM with DEBUG options. See: https://github.com/llvm/llvm-project/blob/9a0ae081047d7088cdecfa86a8c90b721485e418/llvm/lib/CodeGen/MachineScheduler.cpp#L4343

The failing bot was building the LLVM in release mode, therefore, the flag doesn't exist.
Could you revert your change and fix it and reland it please?

vhscampos commented:
Patch reverted while I fix the issue. Thanks for reporting @zeroomega

artagnon commented on llvm/test/CodeGen/ARM/misched-branch-targets.mir:
```diff
@@ -0,0 +1,166 @@
+# RUN: llc -o - -run-pass=machine-scheduler -misched=shuffle %s | FileCheck %s
+# RUN: llc -o - -run-pass=postmisched %s | FileCheck %s
```
Which of these outputs is CHECK'ed? Use `--check-prefix` to disambiguate?

artagnon commented on llvm/test/CodeGen/ARM/misched-branch-targets.mir:
```diff
@@ -0,0 +1,166 @@
+  declare i32 @setjmp(ptr noundef) #1
+  declare void @longjmp(ptr noundef, i32 noundef) #2
+
+  attributes #0 = { "frame-pointer"="all" "target-cpu"="cortex-m55" "target-features"="+armv8.1-m.main" }
+  attributes #1 = { nounwind returns_twice "frame-pointer"="all" "target-cpu"="cortex-m55" "target-features"="+armv8.1-m.main" }
+  attributes #2 = { noreturn nounwind "frame-pointer"="all" "target-cpu"="cortex-m55" "target-features"="+armv8.1-m.main" }
```
These attributes don't seem to be necessary. You can pass the target to llc, and the relevant ones can be added inline.

artagnon commented on llvm/test/CodeGen/ARM/misched-branch-targets.mir:
```diff
@@ -0,0 +1,166 @@
+# RUN: llc -o - -run-pass=machine-scheduler -misched=shuffle %s | FileCheck %s
```
Please generate the file using `update_mir_test_checks.py`.

vhscampos replied:
The script generates many more check lines than necessary, so I prefer to keep the testcase handmade.

artagnon replied:
In other patches that I authored, reviewers strongly suggested that I use the automatic test update scripts, even if it generates more CHECK lines than necessary. In this case, how do I know that the output doesn't diverge in the two llc invocations, within lines you didn't check? If you used `update_mir_test_checks.py`, it would automatically bail out on diverging outputs.

momchil-velikov replied:
Each individual test case is expected to test for specific properties of the compiler output. A test case is not a detector for irrelevant changes. The minimum required amount of check lines means the test won't need to be updated for unrelated changes and in the same time makes is more clear what the test, well, tests.

artagnon commented on llvm/test/CodeGen/ARM/misched-branch-targets.mir:
```diff
@@ -0,0 +1,166 @@
+  target datalayout = "e-m:e-p:32:32-Fi8-i64:64-v128:64:128-a:0:32-n32-S64"
+  target triple = "thumbv8.1m.main-arm-none-eabi"
+
+  define i32 @foo_bti() #0 {
+  entry:
+    ret i32 0
+  }
+
+  define i32 @foo_pac() #0 {
+  entry:
+    ret i32 0
+  }
+
+  define i32 @foo_pacbti() #0 {
+  entry:
+    ret i32 0
+  }
```
These three functions are identical in every way?

momchil-velikov replied:
They are needed only for their names.

artagnon commented on llvm/test/CodeGen/ARM/misched-branch-targets.mir:
```diff
@@ -0,0 +1,166 @@
+    ret i32 0
+  }
+
+  define i32 @foo_setjmp() #0 {
+  entry:
+    ret i32 0
+  if.then:
+    ret i32 0
+  }
```
Isn't this nonsensical LLVM IR?

vhscampos replied:
The LLVM IR for MIR tests can often be nonsensical. Names of functions and basic blocks need to match the ones from the corresponding MIR. Therefore "entry:" and "if.then:" can't be removed.

artagnon reviewed (commented):
I've left a few comments to take into consideration when re-landing.
